File: LeetCode/python3/763.py
class Solution:
    def partitionLabels(self, S: str) -> List[int]:
        start, end, n, pos, res = 0, 0, len(S), [0 for _ in range(26)], []
        for i in range(n):
            pos[ord(S[i]) - ord('a')] = i
        for i in range(n):
            end = max(end, pos[ord(S[i]) - ord('a')])
            if end == i:
                res.append(end - start + 1)
                start = end + 1
                end = start
        return res

File: LeetCode/cpp/763.cpp
class Solution {
public:
    vector<int> partitionLabels(string S) {
        int start = 0, end = 0, n = S.size();
        vector<int> res, pos(26);
        for (int i = 0; i < n; ++i)
            pos[S[i] - 'a'] = i;
        for (int i = 0; i < n; ++i) {
            end = max(end, pos[S[i] - 'a']);
            if (end == i) {
                res.push_back(end - start + 1);
                start = i + 1;
                end = start;
            }
        }
        return res;
    }
};
